include/ldsCtrlEst_h/lds_mpcctrl.h

```diff
@@ -0,0 +1,533 @@
+   */
+  OSQPCscMatrix* from_sparse(const arma::SpMat<data_t>& A);
+
+  /**
+   * @brief     Set the vector for the OSQP solver from an Armadillo vector
+   *
+   * @param     v The vector to convert
+   */
+  OSQPFloat* from_vec(const Vector& v);
+
+  /**
+   * @brief     Calculate the trajectory for the simulation step,
+   *            used when the simulation time step is the same as
+   *            the prior step
+   *
+   * @param     x0  The initial state
+   * @param     u0  The initial control input
+   * @param     xr  The reference trajectory
+   * @param     out Print out step information
+   *
+   * @return    The trajectory for the simulation step
+   */
+  OSQPSolution* fast_update(const Vector& x0, const Vector& u0,
+                            const Matrix& xr, size_t n_sim);
+
+  /**
+   * @brief     Calculate the trajectory for the simulation step
+   *
+   * @param     x0  The initial state
+   * @param     u0  The initial control input
+   * @param     xr  The reference trajectory
+   * @param     out Print out step information
+   *
+   * @return    The trajectory for the simulation step
+   */
+  OSQPSolution* slow_update(const Vector& x0, const Vector& u0,
+                            const Matrix& xr, size_t n_sim);
+
+  /**
+   * @brief     Create an identity matrix with an offset axis
+   * 
+   * @param     n The size of the matrix
+   * @param     k The offset axis
+   * 
+   * @return    The identity matrix with an offset axis
+   */
+  arma::SpMat<data_t> eye_offset(size_t n, int k = -1) {
+    arma::SpMat<data_t> mat(n, n);
+    int start = (k < 0) ? -k : 0;
+    for (int i = start; i < n && i + k < n; i++) {
+      mat(i, i + k) = 1;
+    }
+    return mat;
+  }
+
+  /**
+   * @brief     Create a block diagonal matrix given two input matrices
+   * 
+   * @param     m1 The first matrix
+   * @param     m2 The second matrix
+   * 
+   * @return    The block diagonal matrix
+   */
+  Matrix block_diag(Matrix m1, Matrix m2) {
+    // Calculate the total rows and columns of the block diagonal matrix
+    size_t rows = m1.n_rows + m2.n_rows;
+    size_t cols = m1.n_cols + m2.n_cols;
+
+    // Create the block diagonal matrix
+    Matrix bd = arma::zeros<Matrix>(rows, cols);
+    bd.submat(0, 0, m1.n_rows - 1, m1.n_cols - 1) = m1;
+    bd.submat(m1.n_rows, m1.n_cols, rows - 1, cols - 1) = m2;
+
+    return bd;
+  }
+
+  // void print_osqp_csc_matrix(const OSQPCscMatrix* matrix) {
+  //   std::cout << "[" << matrix->m << ", " << matrix->n << "]:\n";
+
+  //   for (int j = 0, k = 0; j < (matrix->p[matrix->n]); j++) {
+  //     while (k <= matrix->n && matrix->p[k + 1] <= j) k++;
+  //     std::cout << "(" << matrix->i[j]  << ", " << k << ") " << matrix->x[j] << "\n";
+  //   }
+  //   std::cout << std::endl;
+  // }
+
+  void Init() {
+    A_ = sys_.A();
+    B_ = sys_.B();
+    n_ = B_.n_rows;
+    m_ = B_.n_cols;
+    xi_ = arma::zeros<Vector>(n_);
+    t_sim_ = 0;
+  }
+};
+
+// Implement methods
```

Comment:
I'm not super familiar with C++ conventions, but from what I understand these would go in a `.cpp` under `src/`

Reply:
I see. I was just trying to follow the pattern that the other controllers used, because as far as I saw most of the other controllers are implemented like this, but I can do that too!